README.md
# abn-amro-txn

A library for processing bank statements from ABN Amro.

**here be dragons**: This project is mainly a personal experiment for practicing
                     the development workflow for a Haskell project, and to become
                     more familiar with common libraries (e.g. Megaparsec for parsing).
                     **I do not recommend using this for anything serious**.


ChangeLog.md
# Changelog for abn-amro-txn

## Unreleased changes


test/Data/ABNAmro/MT940Spec.hs
{-# options_ghc -Wno-orphans #-}

module Data.ABNAmro.MT940Spec (spec) where

import Control.Applicative (liftA2)

import Data.ByteString (ByteString, pack, unpack)
import Data.ByteString.UTF8 (fromString)
import Data.Foldable
import Data.Time (Day, showGregorian)
import Data.Word

import qualified Text.Megaparsec as MP

import Test.Hspec
import Test.Hspec.Megaparsec
import Test.QuickCheck
import Test.QuickCheck.Instances.ByteString ()
import Test.QuickCheck.Instances.Time ()

import Data.ABNAmro.MT940

-- Useful instances for combining Gens

instance Semigroup a => Semigroup (Gen a) where
    (<>) = liftA2 (<>)

instance Monoid a => Monoid (Gen a) where
    mempty = pure mempty

-- Character sets

nonZeroDigitChars :: Gen Word8
nonZeroDigitChars = elements $ unpack "123456789"

digitChars :: Gen Word8
digitChars = elements $ unpack "0123456789"

alphabeticChars :: Gen Word8
alphabeticChars = elements $ unpack "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

specialChars :: Gen Word8
specialChars = elements $ unpack "\"\\$%&()*+-.,/;:<= "

eol :: Gen ByteString
eol = elements ["\n", "\r\n"]

alphanumericChars :: Gen Word8
alphanumericChars = frequency [
      (75, alphabeticChars)
    , (20, digitChars)
    , (5, specialChars)
    ]

-- Basic value definitions

value :: Gen Word8-> Int -> Int -> Gen ByteString
value chars minSize maxSize = pack <$> (size >>= flip vectorOf chars)
    where size = choose (minSize, maxSize)

digits :: Int -> Gen ByteString
digits = value digitChars 1

digits' :: Int -> Gen ByteString
digits' n = value digitChars n n

nonZeroDigits' :: Int -> Gen ByteString
nonZeroDigits' n = value nonZeroDigitChars n n

alphabetic' :: Int -> Gen ByteString
alphabetic' n = value alphabeticChars n n

alphanumeric :: Int -> Gen ByteString
alphanumeric = value alphanumericChars 1

-- Combinators

optional :: Gen ByteString -> Gen ByteString
optional x = oneof [x, pure ""]

between :: (Int, Int) -> Gen ByteString -> Gen ByteString
between c g = do
    n <- choose c
    fold $ replicate n g

-- Common fields

tag :: ByteString -> Gen ByteString
tag x = pure $ ":" <> x <> ":"

transactionSide :: Gen ByteString
transactionSide = elements ["C", "D"]

extendedTransactionSide :: Gen ByteString
extendedTransactionSide = elements ["C", "D", "RC", "RD"]

fullDate :: Gen ByteString
fullDate = transform <$> (arbitrary :: Gen Day)
    where
        transform = fromString . drop 2 . filter (/= '-') . showGregorian 

condensedDate :: Gen ByteString
condensedDate = transform <$> (arbitrary :: Gen Day)
    where
        transform = fromString . drop 4 . filter (/= '-') . showGregorian

currency :: Gen ByteString
currency = elements [
      "EUR"
    , "USD"
    , "CAD"
    , "GBP"
    -- TODO: add the rest of the ISO currnct codes
    ]

amount :: Gen ByteString
amount = do
    let maxWidth = 15
        maxDigits = maxWidth - 1  -- because we *must* have a decimal point
        maxDecimalPlaces = 2
        decimalPoint = pure ","
        integerPart w = nonZeroDigits' 1 <> digits' (w - 1)
        decimalPart w
            | w == 0 = pure ""
            | otherwise = digits' (w - 1) <> nonZeroDigits' 1
    dw <- choose (0, maxDecimalPlaces)
    mw <- choose (1, maxDigits - dw)
    integerPart mw <> decimalPoint <> decimalPart dw

transactionTypeCode :: Gen ByteString
transactionTypeCode = alphabetic' 1 <> digits' 3


-- Fields

transactionReferenceNumber :: Gen ByteString
transactionReferenceNumber =
    tag "20"
    -- This should be (digits' 8 <> digits' 8), but for some reason all the
    -- MT940 files obtained from ABN Amro just have the string "ABN AMRO BANK NV"
    <> pure "ABN AMRO BANK NV"
    <> eol

relatedReference :: Gen ByteString
relatedReference = pure ""  -- not used in ABN Amro Access Online

accountNumber :: Gen ByteString
accountNumber =
    tag "25" <> alphanumeric 35 <> eol

accountStatementID :: Gen ByteString
accountStatementID =
       tag "28"
    -- This format is specific to accounts in the Netherlands
    <> (fromString . show <$> choose (1, 366 :: Int))  -- day number within current year
    <> pure "01"  -- Run number
    <> optional (pure "/" <> digits 5)  -- Sub-message number
    <> eol

initialBookBalance :: Gen ByteString
initialBookBalance =
       tag "60F"
    <> transactionSide
    <> fullDate
    <> currency
    <> amount
    <> eol

accountStatementTransaction :: Gen ByteString
accountStatementTransaction =
       tag "61"
    <> fullDate
    <> optional condensedDate
    <> extendedTransactionSide
    <> amount
    <> transactionTypeCode
    -- Account holder reference; most of the time this is just NONREF
    <> frequency [
         (80, pure "NONREF")
       , (20, alphanumeric 16)
       ]
    -- Bank reference
    <> optional (pure "//" <> alphanumeric 14)
    -- Transaction information
    <> optional (eol <> pure "/EC-COMPLIANT/")
    <> eol

transactionInformation :: Gen ByteString
transactionInformation =
       tag "86"
    <> between (1, 6) (alphanumeric 65 <> eol)


transaction :: Gen ByteString
transaction = accountStatementTransaction <> optional transactionInformation


finalBookBalance :: Gen ByteString
finalBookBalance =
       tag "62F"
    <> transactionSide
    <> fullDate
    <> currency
    <> amount
    <> eol

valueBalance :: Gen ByteString
valueBalance =
       tag "64"
    <> transactionSide
    <> fullDate
    <> currency
    <> amount
    <> eol

futureValueBalance :: Gen ByteString
futureValueBalance =
       tag "65"
    <> transactionSide
    <> fullDate
    <> currency
    <> amount
    <> eol

messageInformation :: Gen ByteString
messageInformation = pure ""  -- not used in ABN Amro Access Online

-- full message

-- MT940 OfficeNet (exported by ABN Amro Access Online)
-- uses envelope type 2

envelopeHeader :: Gen ByteString
envelopeHeader =
       pure "ABNANL2A" <> eol
    <> pure "940" <> eol
    <> pure "ABNANL2A" <> eol

envelopeTrailer :: Gen ByteString
envelopeTrailer =
    pure "-" <> eol

mt940Message :: Gen ByteString
mt940Message =
    envelopeHeader
        <> transactionReferenceNumber
        <> optional relatedReference
        <> accountNumber
        <> accountStatementID
        <> initialBookBalance
        <> between (0, 100) (
               accountStatementTransaction
            <> transactionInformation
        )
        <> finalBookBalance
        <> optional valueBalance
        <> between (0, 100) futureValueBalance
        <> messageInformation
    <> envelopeTrailer

-- Specs

newtype Any a = Any (Gen ByteString)

canParse :: forall a. (Show a, Field a) => Any a -> SpecWith (Arg Property)
canParse (Any gen) =
    let isParsedBy p = shouldSucceedOn (MP.parse p "")
    in it "can parse anything" $ forAll gen $ isParsedBy (parser :: Parser a)

spec :: Spec
spec = do
    -- Common fields
    describe "TxnSide" $
        canParse (Any transactionSide :: Any TxnSide)
    describe "Currency" $
        canParse (Any currency :: Any Currency)
    describe "Date" $
        canParse (Any fullDate :: Any Date)
    describe "DateNoYear" $
        canParse (Any condensedDate :: Any DateNoYear)
    describe "Amount" $
        canParse (Any amount :: Any Amount)
    describe "TransactionType" $
        canParse (Any transactionTypeCode :: Any TransactionType)
    -- Composite fields
    describe "AccountNumber" $
        canParse (Any accountNumber :: Any AccountNumber)
    describe "StatementID" $
        canParse (Any accountStatementID :: Any StatementID)
    describe "Transaction" $
        canParse (Any transaction :: Any Transaction)
    -- total
    describe "MT940Message" $
        canParse (Any mt940Message :: Any MT940Message)


test/Data/ABNAmro/MT940/AtomsSpec.hs
{-# options_ghc -Wno-orphans #-}

module Data.ABNAmro.MT940.AtomsSpec (spec) where

import Data.Decimal
import Data.Time (Day, fromGregorian)

import Test.Hspec
import Test.Hspec.Megaparsec
import Test.QuickCheck
import Test.QuickCheck.Arbitrary.ADT

import Data.ABNAmro.MT940.Atoms


-- Arbitrary instances for testing
-- TODO: find some way to remove this redundancy

instance Arbitrary TxnSide where
    arbitrary = genericArbitrary

instance Arbitrary BalanceKind where
    arbitrary = genericArbitrary

instance Arbitrary Currency where
    arbitrary = genericArbitrary

instance Arbitrary Day where
    arbitrary = fromGregorian <$> years <*> months <*> days
        where
            -- The format itself encodes years as 2 numbers, so implicitly
            -- it can only handle years in [2000, 2099].
            years = choose (2000, 2099)
            -- fromGregorian makes sure that we get a valid day/month combination
            months = arbitrary
            days = arbitrary

instance Arbitrary Date where
    arbitrary = genericArbitrary

instance Arbitrary DateNoYear where
    arbitrary = genericArbitrary


instance Arbitrary Decimal where
    arbitrary = Decimal <$> arbitrary <*> nonNegative
        where
            nonNegative :: Gen Integer
            nonNegative = fmap getNonNegative (arbitrary :: Gen (NonNegative Integer))

instance Arbitrary Amount where
    arbitrary = genericArbitrary
    

roundtrip :: Field a => a -> Either ParseError a
roundtrip = parse . serialize

spec :: Spec
spec = do
    describe "TxnSide" $
        it "parsing roundtrips with serializing" $
            property $ \(x :: TxnSide) -> roundtrip x `shouldParse`  x

    describe "BalanceKind" $
        it "parsing roundtrips with serializing" $
            property $ \(x :: BalanceKind) -> roundtrip x `shouldParse` x

    describe "Currency" $
        it "parsing roundtrips with serializing" $
            property $ \(x :: Currency) -> roundtrip x `shouldParse` x

    describe "Date" $
        it "parsing roundtrips with serializing" $
            property $ \(x :: Date) -> roundtrip x `shouldParse` x

    describe "DateNoYear" $
        it "parsing roundtrips with serializing" $
            property $ \(x :: DateNoYear) -> roundtrip x `shouldParse` x

    describe "Amount" $
        it "parsing roundtrips with serializing" $
            property $ \(x :: Amount) -> roundtrip x `shouldBe` Right x

src/Data/ABNAmro/MT940/Atoms.hs
module Data.ABNAmro.MT940.Atoms (
      Parser
    , ParseError
    , Serialized
    , Field
    , parser
    , serializer
    , parse
    , serialize
    , serialized
    , read'
    , read''
    , TxnSide
    , BalanceKind
    , Currency
    , Date
    , DateNoYear
    , Amount
) where

import Control.Arrow ((>>>))

import Data.ByteString (ByteString, pack)
import Data.ByteString.UTF8 (fromString, toString)
import Data.Coerce (coerce)
import Data.Decimal
import Data.Ratio ((%))
import Data.Time (Day, fromGregorian, showGregorian)
import Data.Void (Void)
import Data.Word (Word8)

import GHC.Generics

import Text.Megaparsec hiding (Token, ParseError, parse)
import Text.Megaparsec.Byte
import qualified Text.Megaparsec as MP

type Token = Word8
type Serialized = ByteString
type Parser = Parsec Void Serialized
type Serializer a = a -> Serialized
type ParseError = ParseErrorBundle Serialized Void

serialized :: String -> Serialized
serialized = fromString

read' :: Read a => Serialized -> a
read' = read . toString

read'' :: Read a => [Token] -> a
read'' = read' . pack

class Field a where
    serializer :: Serializer a
    parser :: Parser a

parse :: Field a => Serialized -> Either ParseError a
parse = MP.parse parser ""

serialize :: Field a => a -> Serialized
serialize = serializer


-- Atom parsers

-- TODO: use some template haskell to generate the parsers and
--       serializers for these simple datatypes where string
--       literals correspond 1-1 with type constructors

data TxnSide
    = Credit
    | Debit
    | ReverseCredit
    | ReverseDebit
    deriving (Eq, Generic, Show)

instance Field TxnSide where
    parser = choice [
          Credit        <$ chunk "C"
        , Debit         <$ chunk "D"
        , ReverseCredit <$ chunk "RC"
        , ReverseDebit  <$ chunk "RD"
        ]
    serializer t = case t of
        Credit        -> "C"
        Debit         -> "D"
        ReverseCredit -> "RC"
        ReverseDebit  -> "RD"


data BalanceKind
    = Final
    | Intermediate
    deriving (Eq, Generic, Show)

instance Field BalanceKind where
    parser = choice [
          Final        <$ chunk "F"
        , Intermediate <$ chunk "M"
        ]
    serializer k = case k of
        Final        -> "F"
        Intermediate -> "M"


data Currency
    = EUR
    | GBP
    | USD
    | CAD
    deriving (Eq, Generic, Show)

instance Field Currency where
    parser = choice [
          EUR <$ chunk "EUR"
        , GBP <$ chunk "GBP"
        , USD <$ chunk "USD"
        , CAD <$ chunk "CAD"
        ]
    -- The serialized form is the constructor name
    serializer = show >>> serialized


datePart :: Read a => Parser a
datePart = read'' <$> count 2 digitChar

newtype Date = Date Day deriving (Eq, Generic, Show)
instance Field Date where
    parser = Date <$> (fromGregorian <$> year <*> month <*> day)
        where
            year = fmap (2000 +) datePart
            month = datePart
            day = datePart
    serializer = coerce >>> showGregorian >>> filter (/= '-') >>> drop 2 >>> serialized

newtype DateNoYear = DateNoYear Day deriving (Eq, Generic, Show)
instance Field DateNoYear where
    parser = DateNoYear <$> (fromGregorian 0 <$> month <*> day)
        where
            month = datePart
            day = datePart
    serializer = coerce >>> showGregorian >>> filter (/= '-') >>> drop 4 >>> serialized


newtype Amount = Amount Decimal deriving (Eq, Generic, Show)
instance Field Amount where
    parser = do
        let toIntegerPart = (% 1) . read''
            toDecimalPart p = case p of
                Nothing -> 0
                Just x -> read'' x % (10 ^ length x)
        integerPart <- toIntegerPart <$> some digitChar
        chunk ","
        decimalPart <- toDecimalPart <$> optional (some digitChar)
        pure . Amount . fromRational $ integerPart + decimalPart
    serializer
        = (coerce :: (Amount -> Decimal))  -- We need to give the type checker a hand
        >>> show >>> replace '.' ',' >>> ensureDecimal >>> serialized
        where
            replace x y = map (\c -> if c == x then y else c)
            ensureDecimal s
                | ',' `elem` s = s
                | otherwise    = s <> ","

src/Data/ABNAmro/MT940.hs
module Data.ABNAmro.MT940 (
      Parser
    , ParseError
    , Serialized
    , Field
    , parser
    , parse
    , read'
    , read''
    -- basic values
    , TxnSide
    , BalanceKind
    , Currency
    , Date
    , DateNoYear
    , Amount
    , TransactionType
    -- composite values
    , AccountNumber
    , StatementID
    , Transaction
    , MT940Message
) where

import Control.Monad (void)

import Data.ByteString (ByteString, pack, unpack, concat)
import Data.ByteString.UTF8 (toString)
import Data.Decimal
import Data.Time (Day, fromGregorian)
import Data.Void (Void)
import Data.Word (Word8)

import Text.Megaparsec hiding (Token, ParseError, parse, between)
import Text.Megaparsec.Byte hiding (alphaNumChar)
import qualified Text.Megaparsec as MP

type Token = Word8
type Serialized = ByteString
type Parser = Parsec Void Serialized
type ParseError = ParseErrorBundle Serialized Void

read' :: Read a => Serialized -> a
read' = read . toString

read'' :: Read a => [Token] -> a
read'' = read' . pack

class Field a where
    parser :: Parser a

parse :: Field a => Serialized -> Either ParseError a
parse = MP.parse parser ""

-- Basic value definitions

-- TODO: use some template haskell to generate the parsers and
--       serializers for these simple datatypes where string
--       literals correspond 1-1 with type constructors

alphabeticChar :: Parser Token
alphabeticChar = upperChar

specialChar :: Parser Token
specialChar = oneOf $ unpack "\"\\$%&()*+-.,/;:<= "

-- | The subset of special ASCII characters that are allowed in MT940 messages
alphanumericChar :: Parser Token
alphanumericChar = upperChar <|> digitChar <|> specialChar

tag :: Serialized -> Parser ()
tag t = void . chunk $ ":" <> t <> ":"

anyTag :: Parser ()
anyTag = void $ chunk ":" *> count 2 digitChar *> optional upperChar *> chunk ":"


data TxnSide
    = Credit
    | Debit
    | ReverseCredit
    | ReverseDebit
    deriving (Eq, Show)

instance Field TxnSide where
    parser = choice [
          Credit        <$ chunk "C"
        , Debit         <$ chunk "D"
        , ReverseCredit <$ chunk "RC"
        , ReverseDebit  <$ chunk "RD"
        ]


data BalanceKind
    = Final
    | Intermediate
    deriving (Eq, Show)

instance Field BalanceKind where
    parser = choice [
          Final        <$ chunk "F"
        , Intermediate <$ chunk "M"
        ]

data Currency
    = EUR
    | GBP
    | USD
    | CAD
    deriving (Eq, Show)

instance Field Currency where
    parser = choice [
          EUR <$ chunk "EUR"
        , GBP <$ chunk "GBP"
        , USD <$ chunk "USD"
        , CAD <$ chunk "CAD"
        ]


datePart :: Read a => Parser a
datePart = read'' <$> count 2 digitChar

newtype Date = Date Day deriving (Eq, Show)
instance Field Date where
    parser = Date <$> (fromGregorian <$> year <*> month <*> day)
        where
            year = fmap (2000 +) datePart
            month = datePart
            day = datePart

newtype DateNoYear = DateNoYear Day deriving (Eq, Show)
instance Field DateNoYear where
    parser = DateNoYear <$> (fromGregorian 0 <$> month <*> day)
        where
            month = datePart
            day = datePart


newtype Amount = Amount Decimal deriving (Eq, Show)
instance Field Amount where
    parser = do
        integerPart <- pack <$> some digitChar
        chunk ","
        decimalPart <- pack <$> many digitChar
        let a = integerPart
                <> case decimalPart of
                    "" -> ""
                    _ -> "." <> decimalPart
        pure $ Amount $ read' a

data TransactionType = TransactionType Token Int deriving (Eq, Show)
instance Field TransactionType where
    parser =
        TransactionType
            <$> (head <$> count 1 alphabeticChar)
            <*> (read'' <$> count 3 digitChar)

-- Composite values


newtype AccountNumber = AccountNumber ByteString deriving (Eq, Show)
instance Field AccountNumber where
    parser = do
        tag "25"
        acct <- AccountNumber . pack <$> count' 1 35 alphanumericChar
        eol
        pure acct

data StatementID = StatementID Int Int (Maybe Int) deriving (Eq, Show)
instance Field StatementID where
    parser = do
        tag "28"
        s <- count' 3 5 digitChar
        let (sn, rn) = splitAt (length s - 2) s
            statementNumber = read'' sn
            runNumber = read'' rn
        subMessageNumber <- fmap read'' <$> optional (chunk "/" *> count' 1 5 digitChar)
        eol
        pure $ StatementID statementNumber runNumber subMessageNumber


data Transaction =
    Transaction TransactionType Date TxnSide Amount ByteString (Maybe ByteString)
    deriving (Eq, Show)
instance Field Transaction where
    parser = do
        tag "61"
        date <- parser :: Parser Date
        optional $ count 4 digitChar  -- Condensed date; we just ignore
        txnSide <- parser :: Parser TxnSide
        amount <- parser :: Parser Amount
        typeCode <- parser :: Parser TransactionType
        ref <- pack <$> count' 1 16 (notFollowedBy (chunk "//") *> alphanumericChar)
        let line = pack <$> many alphanumericChar <* eol
        line  -- discard remainder of line (including abn reference)
        optional $ notFollowedBy anyTag *> line  -- remove transaction information
        info <- optional $ (:) <$> (tag "86" *> line) <*> many (notFollowedBy anyTag *> line)
        pure $ Transaction typeCode date txnSide amount ref (Data.ByteString.concat <$> info)


data Balance = Balance TxnSide Date Currency Amount deriving (Eq, Show)
instance Field Balance where
    parser = Balance <$> parser <*> parser <*> parser <*> parser <* eol


data MT940Message =
    MT940Message StatementID AccountNumber Balance Balance [Transaction]
    deriving (Eq, Show)
instance Field MT940Message where
    parser = do
        envelopeHeader
        -- For some reason the MT940 messages just have this string in the
        -- "transaction reference" field
        tag "20" *> chunk "ABN AMRO BANK NV" *> eol
        accountNumber <- parser
        statementID <- parser
        initialBalance <- tag "60F" *> parser
        transactions <- count' 0 100 parser
        finalBalance <- tag "62F" *> parser
        -- ignore the "value" balance and "future value" balance
        optional $ tag "64" *> (parser :: Parser Balance)
        count' 0 100 $ tag "65" *> (parser :: Parser Balance)
        envelopeTrailer
        pure $ MT940Message statementID accountNumber initialBalance finalBalance transactions
        where
            envelopeHeader =
                   chunk "ABNANL2A" *> eol
                *> chunk "940" *> eol
                *> chunk "ABNANL2A" *> eol
            envelopeTrailer = chunk "-" *> eol
